File: classworks/ege/27ege/data/75/27-75.pas
##
Assign(input, '27-2022b.txt');
var N := ReadINteger;
var K := 43;
var tailSum := |0bi| + |-1bi|*(K-1);
var tailLen := |0|*K;
var (maxSum, minLen) := (0bi, 10**10);
var totalSum := 0bi;
for var i:=0 to N-1 do begin
  var x := ReadInteger;
  totalSum += x;
  if totalSum < 0 then
    Println( totalSum );
  var r := integer(totalSum mod K);
  if tailSum[r] <> -1 then begin
    var curSum := totalSum - tailSum[r];
    var curLen := i - tailLen[r] + 1;
    if (curSum > maxSum) or 
       ((curSum = maxSum) and (curLen < minLen)) then begin
      maxSum := curSum;
      minLen := curLen;
    end
  end else begin
    tailSum[r] := totalSum;
    tailLen[r] := i+1;
  end
end;
print( minLen );

File: classworks/ege/27ege/data/75/27-75.cpp
#include<iostream>
#include<fstream>
using namespace std;

int main()
{
  const int K = 43;	
  ifstream F("27-2022b.txt");
  int N;
  F >> N;
  long long tailSum[K] = {0};
  for( int i=1;i<K;i++) tailSum[i] = -1;
  int tailLen[K] = {0};
  long long maxSum = 0;
  int minLen = 1000000000;
  long long totalSum = 0;
  for( int i=0; i<N; i++ ) {
    int x;
    F >> x;
    totalSum += x;
    int r = totalSum % K;
    if( tailSum[r] != -1 ) {
      int curSum = totalSum - tailSum[r];
      int curLen = i - tailLen[r] + 1;
      if( (curSum > maxSum) or 
          ((curSum == maxSum) and (curLen < minLen)) ) {
         maxSum = curSum;
         minLen = curLen;
         } 
       } 
     else {
       tailSum[r] = totalSum;
       tailLen[r] = i+1;
	      }
  }
cout << minLen;
}
